ff/src/fields/arithmetic.rs

```diff
@@ -120,8 +121,11 @@ macro_rules! impl_field_square_in_place {
             }
             r[$limbs * 2 - 1] = r[$limbs * 2 - 2] >> 63;
             for i in 0..$limbs {
-                r[$limbs * 2 - 2 - i] =
-                    (r[$limbs * 2 - 2 - i] << 1) | (r[$limbs * 2 - 3 - i] >> 63);
+                let subtractor = (2 * ($limbs - 1usize))
+                    .checked_sub(i + 1)
+                    .map(|index| r[index])
+                    .unwrap_or(0);
```

Comment:
Required to avoid underflowing. I have confirmed locally that it doesn't affect performance.

Reply:
Can a comment be added for the arithmetic operations these lines are doing / intending to do? (Looks like propagating a carry bit?)

I agree this matches what was previously implemented.

Reply:
I'm not too sure myself, but I added a comment explaining what is computed.

Reply:
Ok cool!

Kinda worrisome if neither of us understand the code, we should file an issue to decipher & document whats going on these functions 😅

Reply:
Agreed, @jon-chuang should have the best state, I believe?